clang/lib/Sema/Sema.cpp

```diff
@@ -1118,6 +1118,10 @@ void Sema::ActOnEndOfTranslationUnitFragment(TUFragmentKind Kind) {
   }
 
   if (getLangOpts().SYCLIsDevice) {
+    // Check and propagate device function attributes.
+    MarkDevices();
```

Comment:
Both MarkDevices and ProcessFreeFunctions go over the list of device functions using 
`for (Decl *D : syclDeviceDecls()) {}`
At the beginning of MarkDevices this list is created using DeviceFunctionTracker. I didn't want to traverse the whole CFG twice to generate the device function list.
But, there may be another way. The call to DeviceFunctionTracker could be moved from MarkDevices to ProcessFreeFunctions and then the position of MarkDevices could stay where it was. Let me try that.

Reply:
I was curious why you're using syclDeviceDecls() instead of processing the functions with the attributes as you see them (like we currently do for functions marked with sycl_kernel)

Reply:
I recognize free functions using SYCLAddIRAttributesFunctionAttr>(...) and those attributes get added to the function by MarkDevices via PropagateAndDiagnoseDeviceAttr.

Reply:
What do you mean by gets added to the function? Attributes are added to functions using normal Clang attribute handling. Unless you're talking about attribute propagation of some kind (which happens in `MarkDevices`), the function will the attribute applied to in in `SemaDeclAttr.cpp` (See `handleSYCLAddIRAttributesFunctionAttr`)

Reply:
OK, I can create a CallGraph in ProcessFunctions, traverse all functions and find the ones marked with the free function attributes. That allows the the position of MarkDevices to stay as it was and some tests to remain unchanged.

Reply:
No you don't need to do this. I'm ok with MarkDevices being modified if you need the call graph. It is better to reuse as you have done if you actually do need the call graph. I'm trying to understand why you need a call graph though. If you handle functions when you parse it, you shouldn't need the call graph right?

Reply:
Yes, if an individual function is tested for being a free function and then codegen'd then the CG and searching won't be needed. Im trying that approach.